validator/client/beacon-api/status.go

```diff
@@ -11,69 +11,158 @@ import (
 )
 
 func (c *beaconApiValidatorClient) validatorStatus(in *ethpb.ValidatorStatusRequest) (*ethpb.ValidatorStatusResponse, error) {
-	stringPubKey := hexutil.Encode(in.PublicKey)
+	_, _, validatorsStatusResponse, err := c.getValidatorsStatusResponse([][]byte{in.PublicKey}, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get validator status response")
+	}
 
-	resp := &ethpb.ValidatorStatusResponse{
-		Status:          ethpb.ValidatorStatus_UNKNOWN_STATUS,
-		ActivationEpoch: params.BeaconConfig().FarFutureEpoch,
+	if len(validatorsStatusResponse) != 1 {
+		return nil, errors.New("number of validator status response not expected")
 	}
 
-	stateValidator, err := c.getStateValidators([]string{stringPubKey}, nil)
+	validatorStatusResponse := validatorsStatusResponse[0]
+
+	return validatorStatusResponse, nil
+}
+
+func (c *beaconApiValidatorClient) multipleValidatorStatus(in *ethpb.MultipleValidatorStatusRequest) (*ethpb.MultipleValidatorStatusResponse, error) {
+	publicKeys, indices, statuses, err := c.getValidatorsStatusResponse(in.PublicKeys, in.Indices)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get state validator")
+		return nil, errors.Wrap(err, "failed to get validators status response")
 	}
 
-	// If no data, the validator is in unknown status
-	if len(stateValidator.Data) == 0 {
-		return resp, nil
-	}
+	return &ethpb.MultipleValidatorStatusResponse{
+		PublicKeys: publicKeys,
+		Indices:    indices,
+		Statuses:   statuses,
+	}, nil
+}
 
-	validatorContainer := stateValidator.Data[0]
+func (c *beaconApiValidatorClient) getValidatorsStatusResponse(inPubKeys [][]byte, inIndexes []int64) (
+	[][]byte,
+	[]types.ValidatorIndex,
+	[]*ethpb.ValidatorStatusResponse,
+	error,
+) {
+	totalLen := len(inPubKeys) + len(inIndexes)
 
-	// Set Status
-	status, ok := beaconAPITogRPCValidatorStatus[validatorContainer.Status]
-	if !ok {
-		return nil, errors.New("invalid validator status: " + validatorContainer.Status)
-	}
+	outPubKeys := make([][]byte, totalLen)
+	outIndexes := make([]types.ValidatorIndex, totalLen)
+	validatorsStatus := make([]*ethpb.ValidatorStatusResponse, totalLen)
```

Comment:
Let's move these variables right before the `for` loop because that's where you start using these them.